aws/monitorinstances.go

```diff
@@ -0,0 +1,308 @@
+	InstanceInfo struct {
+		Account    string        `json:"account" type:"string" index:"not_analyzed"`
+		Service    string        `json:"service" type:"string" index:"not_analyzed"`
+		ID         string        `json:"id" type:"string" index:"not_analyzed"`
+		Region     string        `json:"region" type:"string" index:"not_analyzed"`
+		Hour       time.Time     `json:"hour" type:"date" format:"strict_date_optional_time||epoch_millis"`
+		CPUAverage float64       `json:"cpuaverage" type:"double"`
+		CPUPeak    float64       `json:"cpupeak" type:"double"`
+		KeyPair    string        `json:"keypair" type:"string" index:"not_analyzed"`
+		Type       string        `json:"type" type:"string" index:"not_analyzed"`
+		Tags       []InstanceTag `json:"tags" type:"nested"`
+	}
+)
+
+// merge function from https://blog.golang.org/pipelines#TOC_4
+// It allows to merge many chans to one.
+func merge(cs ...<-chan InstanceInfo) <-chan InstanceInfo {
+	var wg sync.WaitGroup
+	out := make(chan InstanceInfo)
+
+	// Start an output goroutine for each input channel in cs. The output
+	// copies values from c to out until c is closed, then calls wg.Done.
+	output := func(c <-chan InstanceInfo) {
+		for n := range c {
+			out <- n
+		}
+		wg.Done()
+	}
+	wg.Add(len(cs))
+	for _, c := range cs {
+		go output(c)
+	}
+
+	// Start a goroutine to close out once all the output goroutines are
+	// done. This must start after the wg.Add call.
+	go func() {
+		wg.Wait()
+		close(out)
+	}()
+	return out
+}
+
+// getAccountID allows to get the AWS Account ID (12 digits) of the
+// credentials given.
+func getAccountID(creds *credentials.Credentials, logger jsonlog.Logger) string {
+	sess := session.Must(session.NewSession(&aws.Config{
+		Credentials: creds,
+		Region:      aws.String(defaultRegion),
+	}))
+	svc := sts.New(sess)
+	res, err := svc.GetCallerIdentity(nil)
+	if err != nil {
+		logger.Error("Get caller identity failed", err)
+		return ""
+	}
+	return aws.StringValue(res.Account)
+}
+
+// getInstanceTag formats an array of *ec2.Tag to an array
+// of InstanceTag.
+func getInstanceTag(tags []*ec2.Tag) []InstanceTag {
+	res := make([]InstanceTag, 0, 5)
+	for _, tag := range tags {
+		res = append(res, InstanceTag{aws.StringValue(tag.Key), aws.StringValue(tag.Value)})
+	}
+	return res
+}
+
+// getInstanceStats gets the CPU average and the CPU peak from CloudWatch
+func getInstanceStats(instanceId string, sess *session.Session, region string, logger jsonlog.Logger) (float64, float64) {
+	now := time.Now()
+	end := time.Date(now.Year(), now.Month(), now.Day(), now.Hour()-1, 0, 0, 0, now.Location())
+	start := time.Date(now.Year(), now.Month(), now.Day(), now.Hour()-2, 0, 0, 0, now.Location())
+	svc := cloudwatch.New(sess)
+	dimensions := []*cloudwatch.Dimension{
+		&cloudwatch.Dimension{
+			Name:  aws.String("InstanceId"),
+			Value: aws.String(instanceId),
+		},
+	}
+	stats, err := svc.GetMetricStatistics(&cloudwatch.GetMetricStatisticsInput{
```

Comment:
Please use `context.Context` to
— provide cancelation of long, blocking operations like network communications
— log stuff (please log more)